programs/drift_vaults/src/state/tokenized_vault_depositor.rs

```diff
@@ -0,0 +1,574 @@
+        self.vault_shares_base = base;
+    }
+
+    fn get_net_deposits(&self) -> i64 {
+        self.net_deposits
+    }
+    fn set_net_deposits(&mut self, amount: i64) {
+        self.net_deposits = amount;
+    }
+
+    fn get_cumulative_profit_share_amount(&self) -> i64 {
+        self.cumulative_profit_share_amount
+    }
+    fn set_cumulative_profit_share_amount(&mut self, amount: i64) {
+        self.cumulative_profit_share_amount = amount;
+    }
+
+    fn get_profit_share_fee_paid(&self) -> u64 {
+        self.profit_share_fee_paid
+    }
+    fn set_profit_share_fee_paid(&mut self, amount: u64) {
+        self.profit_share_fee_paid = amount;
+    }
+
+    fn apply_rebase(
+        &mut self,
+        vault: &mut Vault,
+        vault_protocol: &mut Option<RefMut<VaultProtocol>>,
+        vault_equity: u64,
+    ) -> Result<Option<u128>> {
+        vault.apply_rebase(vault_protocol, vault_equity)?;
+
+        let mut rebase_divisor: Option<u128> = None;
+
+        if vault.shares_base != self.get_vault_shares_base() {
+            validate!(
+                vault.shares_base > self.get_vault_shares_base(),
+                ErrorCode::InvalidVaultRebase,
+                "Rebase expo out of bounds"
+            )?;
+
+            let expo_diff = (vault.shares_base - self.get_vault_shares_base()).cast::<u32>()?;
+
+            rebase_divisor = Some(10_u128.pow(expo_diff));
+
+            msg!(
+                "rebasing vault depositor: base: {} -> {} ",
+                self.get_vault_shares_base(),
+                vault.shares_base,
+            );
+
+            self.set_vault_shares_base(vault.shares_base);
+
+            let old_vault_shares = self.unchecked_vault_shares();
+            let new_vault_shares = old_vault_shares.safe_div(rebase_divisor.unwrap())?;
+
+            msg!(
+                "rebasing vault depositor: shares {} -> {} ",
+                old_vault_shares,
+                new_vault_shares
+            );
+
+            self.update_vault_shares(new_vault_shares, vault)?;
+            self.last_vault_shares = self.get_vault_shares();
+        }
+
+        validate!(
+            self.get_vault_shares_base() == vault.shares_base,
+            ErrorCode::InvalidVaultRebase,
+            "vault depositor shares_base != vault shares_base"
+        )?;
+
+        Ok(rebase_divisor)
+    }
+}
+
+impl TokenizedVaultDepositor {
+    pub fn new(vault: Pubkey, pubkey: Pubkey, mint: Pubkey, now: i64) -> Self {
+        Self {
+            vault,
+            pubkey,
+            mint,
+            vault_shares: 0,
+            last_vault_shares: 0,
+            last_valid_ts: now,
+            net_deposits: 0,
+            total_deposits: 0,
+            total_withdraws: 0,
+            cumulative_profit_share_amount: 0,
+            profit_share_fee_paid: 0,
+            vault_shares_base: 0,
+            bump: 0,
+            padding: [0; 3],
+        }
+    }
+
+    pub fn apply_profit_share(
```

Comment:
are tokenized vaults not allowed to have protocol? surprised there is no vault protocl here